### PR TITLE
Implement async/await approach

### PR DESCRIPTION
**Pull request by Zlata-Zueva**
Imlements async/await approach using `System.Threading.Tasks`. All methods are renamed accordingly, adding `Async` postfix to indicate that they implement async/await approach.

### CODE REVIEW

File: RelationalAI.Test/DatabaseTest.cs [hunk 1]
```diff
@@ -134,29 +135,29 @@ public void DatabaseCloneTest()
...
             Assert.Equal(databaseCloneName, deleteRsp.Name);
         }
 
         public override void Dispose()
         {
             var client = CreateClient();
 
-            try { client.DeleteDatabase(Dbname); } catch {}
-            try { client.DeleteEngineWait(EngineName); } catch {}
+            try { client.DeleteDatabaseAsync(Dbname).Wait(); } catch {}
```
Comment (stanislav-bedunkevich) [inline RelationalAI.Test/DatabaseTest.cs:159]: we probably should at least do Console.Error here in the catch block so it doesn't get lost?

File: RelationalAI.Test/EngineTest.cs [hunk 2]
```diff
@@ -10,35 +11,35 @@ public class EngineTests : UnitTest
         public static string EngineName = $"csharp-sdk-{UUID}";
 
         [Fact]
-        public void EnginetTest()
+        public async Task EngineTest()
         {
             Client client = CreateClient();
 
-            var createRsp = client.CreateEngineWait(EngineName);
+            var createRsp = await  client.CreateEngineWaitAsync(EngineName);
```
Comment (stanislav-bedunkevich) [inline RelationalAI.Test/EngineTest.cs:18]:
```suggestion
            var createRsp = await client.CreateEngineWaitAsync(EngineName);
```

File: RelationalAI.Test/EngineTest.cs [hunk 3]
```diff
@@ -47,7 +48,7 @@ public override void Dispose()
         {
             var client = CreateClient();
 
-            try { client.DeleteEngineWait(EngineName); } catch {}
+            try { client.DeleteEngineWaitAsync(EngineName).Wait(); } catch {}
```
Comment (stanislav-bedunkevich) [inline RelationalAI.Test/EngineTest.cs:51]: same here about COnsole.Err

File: RelationalAI.Test/DatabaseTest.cs (hunk 1 above)
Comment (stanislav-bedunkevich) [reply RelationalAI.Test/DatabaseTest.cs:159]: (just noting things, whether we think it's this PR or not)

File: RelationalAI.Examples/Execute.cs [hunk 4]
```diff
@@ -35,12 +36,12 @@ public static Command GetCommand()
...
         }
 
-        private static void Run(string database, string engine, string command, string profile = "default")
+        private static async Task Run(string database, string engine, string command, string profile = "default")
         {
             Dictionary<string, object> config = Config.Read("", profile);
             Client.Context context = new Client.Context(config);
             Client client = new Client(context);
-            Console.WriteLine(client.Execute(database, engine, command));
+            Console.WriteLine(await client.ExecuteAsync(database, engine, command));
```
Comment (stanislav-bedunkevich) [inline RelationalAI.Examples/Execute.cs:44]:
```suggestion
            Console.WriteLine(await client.ExecuteWaitAsync(database, engine, command));
```

File: RelationalAI/Client.cs [hunk 5]
```diff
@@ -221,11 +223,11 @@ public User CreateUser(string email, List<Role> roles = null)
                 { "email", email },
                 { "roles", uniqueRoles },
             };
-            string resp = this.rest.Post(this.MakeUrl(Client.PathUsers), data) as string;
+            string resp = await this.rest.PostAsync(this.MakeUrl(Client.PathUsers), data) as string;
```
Comment (stanislav-bedunkevich) [inline RelationalAI/Client.cs:226]: we'll need to do `var`ing later too

File: RelationalAI/Client.cs [hunk 6]
```diff
@@ -554,7 +560,7 @@ private string GenLoadJson(string relation)
             var inputs = new Dictionary<string, string>();
             inputs.Add("data", data);
             var source = GenLoadJson(relation);
-            return ExecuteV1(database, engine, source, false, inputs);
+            return ExecuteV1Async(database, engine, source, false, inputs);
```
Comment (stanislav-bedunkevich) [inline RelationalAI/Client.cs:563]: it's a bit strange to see here V1 🤔

File: RelationalAI/Client.cs [hunk 7]
```diff
@@ -654,19 +661,19 @@ private string GenLoadCsv(string relation, CsvOptions options)
             var source = GenLoadCsv(relation, options);
             var inputs = new Dictionary<string, string>();
             inputs.Add("data", data);
-            return ExecuteV1(database, engine, source, false, inputs);
+            return ExecuteV1Async(database, engine, source, false, inputs);
```
Comment (stanislav-bedunkevich) [inline RelationalAI/Client.cs:664]: same here re V1

**Review by stanislav-bedunkevich — APPROVED** [review]
Looks good, thanks!

File: RelationalAI/Client.cs (hunk 6 above)
Comment (stanislav-bedunkevich) [reply RelationalAI/Client.cs:563]: @NRHelmi is there a specific reason we need V1 Execute here for load CSV & JSON?

Comment (NRHelmi) [reply RelationalAI/Client.cs:563]: the sdk still supports v1 execute and one of the usage is loading csv and json data.
we are following a naming convention over all sdks to name the v1 execute/query as `ExecuteV1`

**Comment by larf311** [comment]
Is it idiomatic in C# to append Async to all the method names? I would think the method signature would be sufficient to indicate that it is async.

**Comment by NRHelmi** [comment]
Is there a reason to make all client functions async ? If so let's keep the original functions names otherwise I think `Execute` (maybe `ExecuteV1`) is the only function that needs to be asynchronous.

**Comment by larf311** [comment]
@NRHelmi https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/. It's the idiomatic way of dealing with any network requests.

File: RelationalAI/Client.cs (hunk 6 above)
Comment (stanislav-bedunkevich) [reply RelationalAI/Client.cs:563]: That's fine, just wondering wouldn't it be better to use the newer API if it supports the same?

**Comment by stanislav-bedunkevich** [comment]
Yeah, it is @larf311 @NRHelmi. For example, AWS SDK follows the same approach:
https://github.com/aws/aws-sdk-net/search?p=6&q=Async+Task

File: RelationalAI/Client.cs (hunk 6 above)
Comment (NRHelmi) [reply RelationalAI/Client.cs:563]: v2 is not yet fully implemented API side that's why we still have some similar workarounds in the SDK: for we could use v2 `Execute` to load json and csv data over the wire but this is not how the v2 protocol is designed, I believe the right path is to use multipart forms the same way we are receiving v2 responses but this is not yet supported API side

**Comment by stanislav-bedunkevich** [comment]
> Is there a reason to make all client functions async ? If so let's keep the original functions names otherwise I think `Execute` (maybe `ExecuteV1`) is the only function that needs to be asynchronous.

That is because essentially all SDK methods do end up doing an HTTP request. Now we need to make it async so that thread management could take care of those threads for us and be less burdened. The link Trevor posted above also describes that in more detail.

**Comment by NRHelmi** [comment]
> > Is there a reason to make all client functions async ? If so let's keep the original functions names otherwise I think `Execute` (maybe `ExecuteV1`) is the only function that needs to be asynchronous.
> 
> That is because essentially all SDK methods do end up doing an HTTP request. Now we need to make it async so that thread management could take care of those threads for us and be less burdened. The link Trevor posted above also describes that in more detail.

Fair enough, is it also required to have all async function renamed to `*Async` ?

**Comment by stanislav-bedunkevich** [comment]
> > > Is there a reason to make all client functions async ? If so let's keep the original functions names otherwise I think `Execute` (maybe `ExecuteV1`) is the only function that needs to be asynchronous.
> > 
> > 
> > That is because essentially all SDK methods do end up doing an HTTP request. Now we need to make it async so that thread management could take care of those threads for us and be less burdened. The link Trevor posted above also describes that in more detail.
> 
> Fair enough, is it also required to have all async function renamed to `*Async` ?

Well, it's the common practice in the C# world :) I added a link to the AWS SDK where they also stick to the same approach, even though Java SDK may not have the Async postfixes.

I am not sure we need to have the method names exactly matching across all of the SDKs, as some languages have practices like this one and we'll likely to end up deviating from that. Moreover, we are already kind of, if we consider the various PascalCase, snake_case and camelCase diffs we have in place :)

**Review by NRHelmi — APPROVED** [review]
LGTM thanks :heart:

**Comment by larf311** [comment]
I think the AWS SDK has *Async b/c it also has non-async versions. That being said, the [Google BigQuery API](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.BigQuery.V2/latest/Google.Cloud.BigQuery.V2.BigQueryClient#Google_Cloud_BigQuery_V2_BigQueryClient_Create_System_String_Google_Apis_Auth_OAuth2_GoogleCredential_) also does the same thing (offers both sync & async) so I guess we should do it in case we end up offering both.

**Comment by vilterp** [comment]
Ok, so for the execute functions, we renamed
* `Execute` => `ExecuteWaitAsync`
* `ExecuteAsync` => still `ExecuteAsync`?

Fair enough 🤔 Better than `ExecuteAsyncAsync` 😅

**Comment by stanislav-bedunkevich** [comment]
@larf311 yeah, it seems they both do, right. Although the sync version is usually a wrap around its async version like:
```
        public IDownloadProgress Download(string url, Stream stream)
        {
            return DownloadCoreAsync(url, stream, CancellationToken.None).Result;
        }

        /// <inheritdoc/>
        public async Task<IDownloadProgress> DownloadAsync(string url, Stream stream,
            CancellationToken cancellationToken)
        {
            return await DownloadCoreAsync(url, stream, cancellationToken).ConfigureAwait(false);
        }
```

I agree if we need to provide a sync version later for all of the methods in SDK it would be easier done having the async versions in place (this PR). I am kind of unsure we need that though, as users can write `Client.DoSomethingAsync(...).Result` and make it sync.

Do you have any objections against merging this PR as it is now?

**Review by larf311 — APPROVED** [review]

**Comment by larf311** [comment]
No, looks good. 👍

**Review by stanislav-bedunkevich — APPROVED** [review]

**Pull request merged**